base/setup/lib/utils/partlist.c

```diff
@@ -159,22 +151,17 @@ AssignDriveLetters(
         {
             PartEntry = CONTAINING_RECORD(Entry2, PARTENTRY, ListEntry);
 
-            PartEntry->DriveLetter = 0;
+            if (!PartEntry->Volume)
+                continue;
+            PartEntry->Volume->Info.DriveLetter = 0;
 
             if (PartEntry->IsPartitioned &&
-                !IsContainerPartition(PartEntry->PartitionType))
+                !IsContainerPartition(PartEntry->PartitionType) &&
+                (IsRecognizedPartition(PartEntry->PartitionType) ||
+                 PartEntry->SectorCount.QuadPart != 0LL))
             {
-                ASSERT(PartEntry->PartitionType != PARTITION_ENTRY_UNUSED);
-
-                if (IsRecognizedPartition(PartEntry->PartitionType) ||
-                    PartEntry->SectorCount.QuadPart != 0LL)
-                {
-                    if (Letter <= L'Z')
-                    {
-                        PartEntry->DriveLetter = Letter;
-                        Letter++;
-                    }
-                }
+                if (Letter <= L'Z')
+                    PartEntry->Volume->Info.DriveLetter = Letter++;
```

Comment:
Does this Letter assignment seem to match what the Windows algorithm gives?

Reply:
I just adjusted the code here (for the Volume thing) but actually this doesn't fix any problem of letter assignment, which will be handled later in a separate PR.